iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/compaction/IcebergMajorQueryCompactor.java

```diff
@@ -44,22 +51,69 @@ public boolean run(CompactorContext context) throws IOException, HiveException,
     Map<String, String> tblProperties = context.getTable().getParameters();
     LOG.debug("Initiating compaction for the {} table", compactTableName);
 
-    String compactionQuery = String.format("insert overwrite table %s select * from %<s",
-        compactTableName);
+    HiveConf conf = new HiveConf(context.getConf());
+    String partSpec = context.getCompactionInfo().partName;
+    String compactionQuery;
+
+    if (partSpec == null) {
+      HiveConf.setVar(conf, ConfVars.REWRITE_POLICY, RewritePolicy.ALL_PARTITIONS.name());
+      compactionQuery = String.format("insert overwrite table %s select * from %<s", compactTableName);
+    } else {
+      Table table = IcebergTableUtil.getTable(conf, context.getTable());
+      PartitionData partitionData = DataFiles.data(table.spec(), partSpec);
+      compactionQuery = String.format("insert overwrite table %1$s partition(%2$s) select * from %1$s where %3$s",
+          compactTableName, partDataToSQL(partitionData, partSpec, ","),
+          partDataToSQL(partitionData, partSpec, " and "));
+    }
+
+    SessionState sessionState = setupQueryCompactionSession(conf, context.getCompactionInfo(), tblProperties);
+    sessionState.setConf(conf);
```

Comment:
won't that override the conf setup in setupQueryCompactionSession
````
    Util.disableLlapCaching(conf);
    conf.set(HiveConf.ConfVars.HIVE_QUOTEDID_SUPPORT.varname, "column");
    conf.setBoolVar(HiveConf.ConfVars.HIVE_SERVER2_ENABLE_DOAS, true);
    conf.setBoolVar(HiveConf.ConfVars.HIVE_HDFS_ENCRYPTION_SHIM_CACHE_ON, false);
    Util.overrideConfProps(conf, compactionInfo, tblProperties);
````

Reply:
check MajorQueryCompactor for example, it just clones the conf and passes that to session
````
 HiveConf conf = new HiveConf(hiveConf);
    /*
     * For now, we will group splits on tez so that we end up with all bucket files,
     * with same bucket number in one map task.
     */
    conf.set(HiveConf.ConfVars.SPLIT_GROUPING_MODE.varname, CompactorUtil.COMPACTOR);
````

Reply:
When the cloned conf is passed to the session, and the session is not null, the conf is not set in the session and any changes defined in it get lost. I found it during debugging when trying to understand why in a q-test that has 2 compaction calls the second compaction command is not working.

```
  public static SessionState setUpSessionState(HiveConf conf, String user, boolean doStart) {
    SessionState sessionState = SessionState.get();
    if (sessionState == null) {
      // Note: we assume that workers run on the same threads repeatedly, so we can set up
      //       the session here and it will be reused without explicitly storing in the worker.
      sessionState = new SessionState(conf, user);
      if (doStart) {
        // TODO: Required due to SessionState.getHDFSSessionPath. Why wasn't it required before?
        sessionState.setIsHiveServerQuery(true);
        SessionState.start(sessionState);
      }
      SessionState.setCurrentSessionState(sessionState);
    }
    return sessionState;
  }
```

Reply:
how does it work then for ACID compactions? shouldn't we fix that in setUpSessionState and always update the conf?

Reply:
Yes, I think we should. By looking the at code, I think it behaves the same for ACID compactions, meaning that conf is not updates when using the same session..

Reply:
```` 
else {
      sessionState.setConf(conf);
    }
````

Reply:
Done